### ISSUE TITLE
wrong anti aim detection

### ISSUE THREAD

**Issue by bs321** [BeepIsla/CSGO-Overwatch-Bot#158]
Like the title suggests, anti aim detection is extremely wrong.

**Comment by BeepIsla** [comment]
Aimbot has not been touched if I remember correctly

**Comment by BeepIsla** [comment]
@BlackYuzia any idea about the AntiAim part?

**Comment by BlackYuzia** [comment]
@BeepIsla 

About aimbot part - we just change `death` event to `hurt`.

`hurt` event has been call more times.
Maybe because that we can get more detects from clear players.

About AA part. 
@bs321 can you share demo with me? 

Also. Give a screenshot with current results (detects and verdict)

**Comment by BlackYuzia** [comment]
About warmup false-positive. 
If that true I can return check warmup.

**Comment by BeepIsla** [comment]
Right I forgot about the death event to hurt event change, the hurt event simply triggers *a lot* more often, the normal aimbot detection already had 1-5 or so detections but it wasn't enough to convict, the hurt event will have a lot more of those since it triggers more often. Might need to adjust the default config.

**Comment by BlackYuzia** [comment]
@BeepIsla oh yep.
When I test aimbot detector on legit players they get smaller aimbot detects like clear players. Lol.

If you ask why - they use **triggerbot** a lot of time.

**Comment by bs321** [comment]
@BlackYuzia @BeepIsla http://www.mediafire.com/file/6v50cq43pvz70t0/3.dem/file

run the demo through the bot with steam id 76561198266718297

have a look at the game as well then through the csgo client

**Comment by bs321** [comment]
//False Positive for Aimbot

@BlackYuzia @BeepIsla http://www.mediafire.com/file/o6gb11co458x8ac/5.dem/file

run the demo through the bot with steam id 76561198266718297

have a look at the game as well then through the csgo client

**Comment by bs321** [comment]
Please fix the aimbot detection as well. It's extremely poor right now. False positives everywhere.

**Comment by BeepIsla** [comment]
Fix it yourself or wait. Don't just complain.

**Comment by bs321** [comment]
It's pointless for me to fix it. Your bot is falsely convicting clean players out there.

**Comment by BeepIsla** [comment]
Overwatch cases are viewed by many people before the suspect is banned. A couple of false verdicts are likely not going to have any impact.

**Comment by bs321** [comment]
Sorry if I am being impatient but there's a possibility of someone getting convicted because same demo went to multiple people using OW bot, right?

**Comment by BeepIsla** [comment]
Of course its possible, but false bans are **always** possible with Overwatch. Its Valve's job to make sure there are no false bans and to ban people automating Overwatch or just making the bot not work.

**Comment by bs321** [comment]
Right, thank you for clarifying.

**Comment by BlackYuzia** [comment]
> //False Positive for Anti Aim

Okey, I has been remove check demo if warmup period is enable.
So, I get only 32 (65 - 33) false-positive results.

> ![image](https://user-images.githubusercontent.com/25010528/95209946-9b5ddf80-07f3-11eb-9f40-2ca4eae9ab18.png)
 

But I think this didn't full successfully fix. Kinda TYPO fix or something like this.

@bs321 so, if you `duck` AA detector was think what you are use AA?

**Comment by bs321** [comment]
@BlackYuzia I see. It's not only for warmup. Anytime you duck in game, it's detected as AA.

**Comment by bs321** [comment]
If you see the demo, you will see that person was ducking on box on Mirage looking at A main, so that palace guy cannot see him. That's being detected as AA on the bot.

**Comment by BlackYuzia** [comment]
@bs321 `76561198266718297` didn't use hacks?

**Comment by bs321** [comment]
Play a game and like sit somewhere holding ctrl (duck) and then run the game through the bot. @BlackYuzia

**Comment by BlackYuzia** [comment]
Yep, AA detector is suck big black dick.

I try fix that. Also I try create new method for detect AA.
I check what really detect my AA and sometimes he find some EPIC shit. :eyes:

**Comment by bs321** [comment]
@BlackYuzia Yes. Thank you brother. Too many false positives. Aimbot too, not sure what's wrong since last update but something's isn't right there as well. 
You know how when someone is using AA their legs switch/flick, right? Hope that can help.

**Comment by bs321** [comment]
or something like someone is looking somewhere else, but shoots at target?

**Comment by BlackYuzia** [comment]
> or something like someone is looking somewhere else, but shoots at target?

Already. That checks `aimbot` detector.
But that sometimes can have false-positive.

**Comment by BlackYuzia** [comment]
@BeepIsla `tickEnd(tick)` was return current tick or tick+1?

I can't find some ticks.

**Comment by BeepIsla** [comment]
After everything within a tick has been parsed, it will fire with the current tick.

Some ticks just don't trigger this for whatever reason... Possibly because demos are 32 tick and matches are 64 tick?

**Comment by BlackYuzia** [comment]
For example:

I can't find ticks `45076, 45114, 45284`
But I can find `44973, 45285, 45283`

**Comment by BlackYuzia** [comment]
@BeepIsla so, I need add one tick (value + 1) or del one tick (val - 1)

**Comment by BlackYuzia** [comment]
> Why would you want to do that?

I want check 100% what angles really have in a `normal player` and cheater with AA in some array of ticks.

**Comment by BeepIsla** [comment]
That makes no sense, OnTickEnd already just runs every* tick with the current tick. + or - anything makes no sense.

**Comment by BlackYuzia** [comment]
I was break this game lol
![image](https://user-images.githubusercontent.com/25010528/95222394-1d083a00-0801-11eb-813c-2f09bb1de069.png)

PS: that not AA

**Comment by Musta40** [comment]
I installed latest version on overwatch bot. It says i need to install phynton? Do i really have ti install? Drop a link please from there i can download. Thanks

**Comment by BeepIsla** [comment]
Do **not** hijack other issues, create your own. You only need it for the NodeJS build tool for running in `Steamworks`-Mode which is optional. Everything is explained in the README, read it.

**Comment by BlackYuzia** [comment]
@BeepIsla did you want a joke? :eyes:
So, I think `AA detector` fully suck.
I can't create good method for detect AA.

But I was found what I can check angles of players per tick.
I get success results for ONLY ONE cheater (who really use AA).
All clear players (or cheaters without AA) get 0 or < 20 detects.

### Screenshots

#### Cheater 

> ![image](https://user-images.githubusercontent.com/25010528/95232072-b12bce80-080c-11eb-8124-4a4457408754.png)

#### No AA

> ![image](https://user-images.githubusercontent.com/25010528/95232260-eb956b80-080c-11eb-91c3-d002f46ce32b.png)

Also other guys of this demo (and of demo by @bs321) I get clear results for guys who didn't use AA.

**Comment by Musta40** [comment]
Sry and thanks guys

Regards
             Mantas

**Comment by BlackYuzia** [comment]
But I have some troubles.

1. I check one angle so if cheaters have 2-3 different types of angles I just suck keys.
2. I need more demos for check did this method can detect >30-50% or only ONE GUY with AA :D

**Comment by BlackYuzia** [comment]
So... @BeepIsla @bs321 and any other guys can you share with me demos with Rage or HvH (where cheaters use AA)

I need more tests now.

**Comment by BlackYuzia** [comment]
Oh yep.
A few words about AA detector.

I want totaly delete AA detector.
But add new with name `AnglesCheck` or something like this.

@BeepIsla and yep. **Can you disable/delete current AA detector?**

**Comment by BeepIsla** [comment]
1e102761e81d191a02acdff926d58d009e7ad7cd

**Comment by bs321** [comment]
@BlackYuzia I can try you gotta let me know if what you are trying to detect is legit AA or rage AA

**Comment by BlackYuzia** [comment]
@BeepIsla Can you explain to me how I can add a new Detector? I can't understand how to :eyes: 
Or I should just upload new code in to `AA detector` and you do all the other?

> @BlackYuzia I can try you gotta let me know if what you are trying to detect is legit AA or rage AA

I upload my "new" version in to my repo and will give you link. Wait a bit.

**Comment by BlackYuzia** [comment]
@bs321 you can download [this branch](https://github.com/BlackYuzia/CSGO-Overwatch-Bot/tree/BlackYuzia-experimental-angles-check) or [this file](https://github.com/BlackYuzia/CSGO-Overwatch-Bot/blob/BlackYuzia-experimental-angles-check/detectors/AntiAim.js) and replace old file.

Try run it.
But I would want get demos for test by myown :eyes:

**Comment by BlackYuzia** [comment]
oh, I was forgot. @bs321 also change value of AA detects to 1000+

**Comment by BlackYuzia** [comment]
@BeepIsla I don't know how, why and how this work but this work :D

All **AA Cheaters** what I check really have AA and get `>= 10K detects`.
I think we can wait feedback from @bs321 and if all is good I would create PR with addings a new AA detect method.

Also I has been started work on **legit AA** detects. 
I don't know did that can work the same but I try and test now.

**Comment by bs321** [comment]
@BlackYuzia Give me some time. I will run the demo through the updated version. 
I am having difficulties finding demos in which there was hvh with anti aim and stuff.  

The legit AA, I don’t think it will help. Because in OW demo, you’re locked with one single POV and there is no way to check for legit AA. Even if a human were to inspect it, he wouldn’t know. You need to inspect it in 3rd person to give a verdict and since OW demos don’t allow it, I don’t think it is possible.

**Comment by BlackYuzia** [comment]
If anybody want would get a demo with Legit AA Cheaters I have one of [Demo](https://gofile.io/d/VOCNck) 

1. 76561199068955562
2. 76561199068893162

This both 100% use **Legit AA**.

**Comment by BlackYuzia** [comment]
> The legit AA, I don’t think it will help. Because in OW demo, you’re locked with one single POV and there is no way to check for legit AA. Even if a human were to inspect it, he wouldn’t know. You need to inspect it in 3rd person to give a verdict and since OW demos don’t allow it, I don’t think it is possible.

Yep, but if >50% was use this bot on one guy ... **HE WILL BE BANNED** :eyes:

**Comment by BlackYuzia** [comment]
> I am having difficulties finding demos in which there was hvh with anti aim and stuff.

[You can check demo from here](https://github.com/BeepIsla/CSGO-Overwatch-Bot/issues/158#issuecomment-704482521) if you want.
**I tested on this demo.**

But if you find a new maybe this would be better.

**Comment by bs321** [comment]
@BlackYuzia I agree but then you have to make sure that there are absolutely no false positives. Because then it kills the whole point doing it.

**Comment by BlackYuzia** [comment]
> @BlackYuzia I agree but then you have to make sure that there are absolutely no false positives. Because then it kills the whole point doing it.

I need a bit more demos with `legit AA`.
If you found that share with me pls.

**Comment by bs321** [comment]
> > I am having difficulties finding demos in which there was hvh with anti aim and stuff.
> 
> [You can check demo from here](https://github.com/BeepIsla/CSGO-Overwatch-Bot/issues/158#issuecomment-704482521) if you want.
> **I tested on this demo.**
> 
> But if you find a new maybe this would be better.

Ok I am away from the rig atm. I will get home and test it.

**Comment by bs321** [comment]
> > @BlackYuzia I agree but then you have to make sure that there are absolutely no false positives. Because then it kills the whole point doing it.
> 
> I need a bit more demos with `legit AA`.
> If you found that share with me pls.

Yes let me get back home. Will go through the history of games and help you find it.

**Comment by BlackYuzia** [comment]
@bs321 some note - I think demos with `Rage AA` required too :eyes:

**Comment by bs321** [comment]
@BlackYuzia http://www.mediafire.com/file/h0t7i3vucv9h78v/12.dem/file

Download this demo. Almost everyone is using legit AA namely,  76561197960909053, 76561199079155684, 76561198872552000, 76561198354741622, 76561198064550048 and 76561198180436401. I think this would be a good starting point as you can test these many players in one single game.

**Comment by bs321** [comment]
I will try finding demos with Rage AA, it's difficult to find as older demos have expired and it's rare to find like a rage aa in prime mm with good trust factor.

**Comment by KillaBoi** [comment]
Would it not be easy to just go into a local server, turn on Rage AA or Legit AA or whatever is required and start recording in GOTV mode and give you that instead? That's how I used to test things back when I had my stuff running... It should work and it should be easier than trying to find demos from MM games because it gives you more flexibility to change around and test whatever is needed

**Comment by BlackYuzia** [comment]
@KillaBoi right.

**Comment by BlackYuzia** [comment]
@bs321 @BeepIsla  I try create AnglesChecker for Legit AA too.
**But I think this unstable.**

In tests what I complete I was getting next results:

**- Cheaters with Rage AA** - detects <= 500 / 1000
**- Cheaters with Legit AA** -  I can get two type of results
  - 1 type - detects >= 1000 / 1000
  - 2 type - detects <= 700 / 1000 
(idk why and how but somebody cheaters who really use legit AA didn't get 1000 detects or more) | Verdict - need more tests or change method.
**- Cheaters without AA** - one guy who have cheaters get >= 1000 / 1000 detects
**- Clear Players** - <= 500 / 1000 detects.

I should spend more time for legit AA I think.
Also I need saw more types of rage AA.

**Comment by BlackYuzia** [comment]
@bs321 did you test?

**Comment by bs321** [comment]
I did test using 'CSGO-Overwatch-Bot-BlackYuzia-experimental-angles-check'. It's unable to detect legit AA. The one's in which it detects, the final verdict is negative

**Comment by BlackYuzia** [comment]
Yep. This branch can detect only rage AA.
I didn't say what this branch can detect legit AA :eyes:

But I think I complete `preversion` of legit AA detector.
I push it to the same branch a bit later

**Comment by BlackYuzia** [comment]
@bs321 if you want you can try download this branch again https://github.com/BlackYuzia/CSGO-Overwatch-Bot/tree/BlackYuzia-experimental-angles-check

I has been update config and add legit AA detector (also rename old rage AA detector)
But we should complete more tests for legit and rage detectors of AA.

**Comment by BlackYuzia** [comment]
@bs321 call me If you complete some tests.

**Comment by BlackYuzia** [comment]
LOL.
**If I right we can't detect `Legit AA` with GoTV.**

I check some docs how work legit AA.
In simple - he stop send packets, then he changle angle to -90 ( + 90) then start send packets and set angle back (to prev).
So clients (also gotv too) get second angle like "real" angle. But this angle are fake.

So, we can't get real angle of player. 
In Rage AA we can detect when user look at floor.

**I hope I did't right.**

**Comment by BlackYuzia** [comment]
@bs321 forgot about tests legit AA detector.
It would not work :c 

Only Rage AA Detector. Only **Hard~~core~~code**

**Comment by KillaBoi** [comment]
Rage AA would be fine with most people tbh, we could also pick up when the Legit AA flicks due to a bad connection or high ping because it gives you a ton of detections currently when it happens...

**Comment by BlackYuzia** [comment]
@KillaBoi @bs321 try to use this branch - https://github.com/BlackYuzia/CSGO-Overwatch-Bot/tree/BlackYuzia-lbydelta-checker

I will check some sources how to really works AA.
In my tests I get 1K+ detects of rage AA also legit AA too. But IDK how and **how** all of this works :eyes:

Check pls because I didn't have time for check a lot of demos.

I also get false-posiitve sometimes but verdict didn't say what this user have AA. 
Maybe I forgot something ... but now it is all what I want to say.

Feedback me pls if you do some tests with different players, types of AA and etc.

**Comment by bs321** [comment]
@BlackYuzia I will test it, been a lil busy since past couple of days. There's a bigger issue, false positives on aimbot. It's still detecting legit player as using aimbot.

**Comment by BlackYuzia** [comment]
About aimbot call @BeepIsla 
Maybe we should return player_death event.

**Comment by creeperdude997** [comment]
Just adding to this, using this replay with steam ID 76561199084564494 gives the results below, even though they are looking at the ground and getting kills. Only a 2 aimbot rating.
[OW_replay9.zip](https://github.com/BeepIsla/CSGO-Overwatch-Bot/files/6703098/OW_replay9.zip)
![image](https://user-images.githubusercontent.com/75228915/123124729-d49cb500-d43f-11eb-8217-e8bd7d9d0d87.png)

**Comment by BeepIsla** [comment]
I have adjusted the aimbot detector, it should be *much* better now. Note however that the default `threshold` in the config has changed from `1.0` to `5.0`

**Comment by BlackYuzia** [comment]
@BeepIsla how possible detect legit aimbot?
I mean, did we can do that only with demo?

**Comment by BeepIsla** [comment]
You can detect static movement, aimbots typically move the crosshair in a single straight line which is detectable, some aimbots add a bit of randomness to it though which makes it *a lot* harder, if not impossible, to detect.
